### PR TITLE
feat: Add C codes for bucket sort, heap sort, selection sort, binary search edge

### PR DESCRIPTION
**Pull request by Gonglja**
If this PR is related to coding or code translation, please fill out the checklist and paste the console outputs to the PR.

- [x] I've tested the code and ensured the outputs are the same as the outputs of reference codes.
- [x] I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
- [x] The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).

### CODE REVIEW

File: codes/c/chapter_sorting/selection_sort.c [hunk 1]
```diff
@@ -0,0 +1,43 @@
+/**
+ * File: selection_sort.c
+ * Created Time: 2023-05-31
+ * Author: Gonglja (glj0@outlook.com)
+ */
+
+#include "../utils/common.h"
+
+/* 交换两个元素的值 */
+void swap(int *a, int *b) {
```
Comment (krahets) [inline codes/c/chapter_sorting/selection_sort.c:10]: Recommend implementing the swap operation in the `selectionSort()` function. Please refer to Java codes.

Comment (Gonglja) [reply codes/c/chapter_sorting/selection_sort.c]: OK, updated.

Comment (Gonglja) [reply codes/c/chapter_sorting/selection_sort.c]: The swap function is also used in `selection_sort.cpp`, do you need to change it?

Comment (krahets) [reply codes/c/chapter_sorting/selection_sort.c]: No, swap() is a built-in function in C++ std.

File: codes/c/utils/print_util.h [hunk 2]
```diff
@@ -39,6 +39,27 @@ static void printArray(int arr[], int size) {
     }
 }
 
+/* Print an Array */
+static void printArray1(float arr[], int size) {
```
Comment (krahets) [inline codes/c/utils/print_util.h:43]: printArrayFloat()

File: codes/c/chapter_sorting/bucket_sort.c [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+
+/* Driver Code */
+int main() {
+    // 设输入数据为浮点数，范围为 [0, 1)
+    float nums[ARRAY_SIZE] = {0.49f, 0.96f, 0.82f, 0.09f, 0.57f, 0.43f, 0.91f, 0.75f, 0.15f, 0.37f};
+    bucketSort(nums, ARRAY_SIZE);
+    printf("桶排序完成后 nums = ");
+    printArray1(nums, ARRAY_SIZE);
```
Comment (krahets) [inline codes/c/chapter_sorting/bucket_sort.c:81]: printArrayFloat()

File: codes/c/chapter_sorting/CMakeLists.txt [hunk 4]
```diff
@@ -4,3 +4,6 @@ add_executable(quick_sort quick_sort.c)
 add_executable(counting_sort counting_sort.c)
 add_executable(radix_sort radix_sort.c)
 add_executable(merge_sort merge_sort.c)
+add_executable(head_sort head_sort.c)
```
Comment (krahets) [inline codes/c/chapter_sorting/CMakeLists.txt:7]: heap sort

File: codes/c/chapter_sorting/bucket_sort.c [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        // 输入数据范围 [0, 1)，使用 num * k 映射到索引范围 [0, k-1]
+        int bucket_idx = nums[i] * k;
+        int j = 0;
+        // 如果桶中有数据且数据小于当前值 nums[i], 要将其放到当前桶的后面，相当于 cpp 中的 push_back
+        while (buckets[bucket_idx][j] > 0 && buckets[bucket_idx][j] < nums[i]) {
+            j++;
+        }
+        float temp = nums[i];
+        while (j < ARRAY_SIZE && buckets[bucket_idx][j] > 0) {
+              swap(&temp, &buckets[bucket_idx][j]);
```
Comment (krahets) [inline codes/c/chapter_sorting/bucket_sort.c:47]: indentation

**Comment by krahets** [comment]
Hi @Gonglja ! please check the codes carefully.

File: codes/c/chapter_backtracking/preorder_traversal_i_compact.c [hunk 6]
```diff
@@ -0,0 +1,85 @@
+/**
+ * File: preorder_traversal_i_compact.c
```
Comment (krahets) [inline codes/c/chapter_backtracking/preorder_traversal_i_compact.c:2]: I recommend that create another PR for the chapter backtracking. It is more easily to trace the PR.

**Review by krahets — APPROVED** [review]
Thanks @Gonglja !

**Pull request merged**